tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
-            // if (this.hi < this.maxHi) return -1;
-            if (this.hitEOF) return -1;
-                
-            // slow path -- read another buffer
-            this.seek(this.curr);
-            if (this.curr == this.hi) return -1;
+        // NOTE: single-byte reads are common enough to justify having an optimized procedure for them.
+
+        // Check for EOF
+        if (this.curr >= this.length) {
+            return -1;
         }
-        // Assert.check(this.curr < this.hi);
-        try {
-        	byte res = this.buff[(int)(this.curr - this.lo)];
-        	this.curr++;
-        	return ((int)res) & 0xFF; // convert byte -> int
-        } catch (ArrayIndexOutOfBoundsException e) {
-        	throw new IOException("Read past end of file (increase with setLength)?", e);
+
+        // Read the next byte out of the cache.  Invariant V2 guarantees that this read is in-bounds.
+        int result = Byte.toUnsignedInt(this.buff[(int)(this.curr - this.lo)]);
+        ++this.curr;
+
+        // If we read past the buffer, restore our internal invariants
+        if (this.curr >= this.lo + BuffSz) {
+            seeek(this.curr);
         }
+
+        return result;
     }
-    
-    /* overrides RandomAccessFile.read(byte[]) */
+
+    @Override
     public int read(byte[] b) throws IOException {
         return this.read(b, 0, b.length);
     }
-    
-    /* overrides RandomAccessFile.read(byte[], int, int) */
+
+    @Override
     public int read(byte[] b, int off, int len) throws IOException {
         // Assert.check(!this.closed);
-        if (this.curr == this.hi) {
-            // test for EOF
-            // if (this.hi < this.maxHi) return -1;
-            if (this.hitEOF) return -1;
-                
-            // slow path -- read another buffer
-            this.seek(this.curr);
-            if (this.curr == this.hi) return -1;
+        if (len == 0) {
+            return 0;
         }
-        // Assert.check(this.curr < this.hi);
-        len = Math.min(len, (int)(this.hi - this.curr));
+
+        // Read what we can into "b"
+        long numReadableWithoutSeeking = Math.min(this.lo + BuffSz, this.length) - this.curr;
+
+        // Check for EOF
+        if (numReadableWithoutSeeking <= 0) {
+            return -1;
+        }
+
+        int numToRead = Math.min(len, (int)numReadableWithoutSeeking);
         int buffOff = (int)(this.curr - this.lo);
-        System.arraycopy(this.buff, buffOff, b, off, len);
-        this.curr += len;        
-        return len;
-    }
+        System.arraycopy(this.buff, buffOff, b, off, numToRead);
+        this.curr += numToRead;
 
-    public BigInteger readBigInteger(int size) throws IOException {
-      byte[] b = new byte[size];
-      // Assert.check(this.read(b) == size);
-      return new BigInteger(b);
+        // If we read past the buffer, restore our internal invariants
+        if (this.curr >= this.lo + BuffSz) {
+            seeek(this.curr);
```

Comment:
I am inclined to factor this out into a helper method with a name like `restoreInv2`, since an identical check and `seeek` call appears other places as well. I will make that change soon. (When I do, I will probably keep the call to `seeek` and not `seek`, since it avoids an extra call stack entry.)